### PR TITLE
Add 64 bit variant to dealii.

### PR DESCRIPTION
**Pull request by Rombur**

### CODE REVIEW

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 1]
```diff
@@ -241,6 +243,16 @@ def cmake_args(self):
                 '-DDEAL_II_WITH_OPENCASCADE=OFF'
             ])
 
+        # 64 bit indices
+        if '+64bit' in spec:
+            options.extend([
+                '-DDEAL_II_WITH_64BIT_INDICES=ON'
+            ])
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dealii/package.py]: You can just use:
```python
options.append('-DDEAL_II_WITH_64BIT_INDICES=ON')
```
and it won't need to create a new list.

**Comment by adamjstewart** [comment]
@davydden

**Comment by davydden** [comment]
do we not need, at least for now, disable `petsc` and `trilinos` by doing 
```
 depends_on("petsc+mpi",        when='@8.4.2:+petsc+mpi~64bit')
 depends_on("slepc",            when='@8.4.2:+slepc+petsc+mpi~64bit')
 depends_on("petsc@:3.6.4+mpi", when='@:8.4.1+petsc+mpi~64bit')
 depends_on("slepc@:3.6.3",     when='@:8.4.1+slepc+petsc+mpi~64bit')
 depends_on("trilinos",         when='+trilinos+mpi~64bit')
```
? 

Otherwise I expect `spack install dealii+64bit` to fail to configure.

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 2]
```diff
@@ -241,6 +243,16 @@ def cmake_args(self):
                 '-DDEAL_II_WITH_OPENCASCADE=OFF'
             ])
 
+        # 64 bit indices
+        if '+64bit' in spec:
+            options.extend([
+                '-DDEAL_II_WITH_64BIT_INDICES=ON'
```
Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: i would keep it in a single line by
```
'-DDEAL_II_WITH_64BIT_INDICES=%s' % (
         'ON' if '+64bit' in spec else 'OFF')
```

**Comment by Rombur** [comment]
@davydden `trilinos` supports 64 bit indices by default but you are right about `petsc`. `petsc` needs to be configured using `--with-64-bit-indices` I never install `petsc` so I forgot about it.

**Comment by Rombur** [comment]
done

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 3]
```diff
@@ -107,11 +109,11 @@ class Dealii(CMakePackage):
...
     depends_on("p4est",            when='+p4est+mpi')
-    depends_on("petsc+mpi",        when='@8.4.2:+petsc+mpi')
+    depends_on("petsc+mpi",        when='@8.4.2:+petsc+mpi~64bit')
     depends_on('python',           when='@8.5.0:+python')
-    depends_on("slepc",            when='@8.4.2:+slepc+petsc+mpi')
-    depends_on("petsc@:3.6.4+mpi", when='@:8.4.1+petsc+mpi')
-    depends_on("slepc@:3.6.3",     when='@:8.4.1+slepc+petsc+mpi')
+    depends_on("slepc",            when='@8.4.2:+slepc+petsc+mpi~64bit')
+    depends_on("petsc@:3.6.4+mpi", when='@:8.4.1+petsc+mpi~64bit')
+    depends_on("slepc@:3.6.3",     when='@:8.4.1+slepc+petsc+mpi~64bit')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/dealii/package.py]: This doesn't require `slepc+64bit` or `petsc+64bit` `when='+64bit'`?

Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: ... someone like me https://github.com/LLNL/spack/pull/2655 :smile:

Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: anyway, i would still merge this as-is. Once both are merged, i will adjust deal.II to use petsc+64bit and see if it compiles.

File: var/spack/repos/builtin/packages/dealii/package.py (hunk 2 above)
Comment (tgamblin) [inline var/spack/repos/builtin/packages/dealii/package.py]: Can I advocate for a new idiom here?  You know you can do this, right?

```python
'-DDEAL_II_WITH_64BIT_INDICES=%s' % ('+64bit' in spec)
```

That `str(<some bool>)` resolves to `True` or `False`, and `CMake` understands any case variation of `TRUE`/`FALSE`, [among other things](https://cmake.org/Wiki/CMake/Language_Syntax#CMake_supports_boolean_variables.).  Even things ending in `-NOTFOUND` are false to `CMake`, though that was always a bit too much like a [daily WTF](http://thedailywtf.com/articles/What_Is_Truth_0x3f_) for my taste.

Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: interesting, did not know that.

Comment (tgamblin) [inline var/spack/repos/builtin/packages/dealii/package.py]: I don't mind merging as-is but this could probably shorten up some peoples' code.

Comment (Rombur) [inline var/spack/repos/builtin/packages/dealii/package.py]: Don't merge it. I will update the PR.

**Comment by tgamblin** [comment]
@davydden @Rombur: merging but weren't we going to change this to `int64` instead of `64bit`?  Or does the variant mean something else here?

**Pull request merged**